Add sf-rest inbound endpoint example (#1468)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/salesforce-rest-connector/sf-rest-inbound-endpoint-example.md

# Salesforce Inbound Endpoint Example

> In the real world, the Salesforce streaming Inbound Endpoint allows you to perform various operations on Salesforce streaming data via WSO2 EI.

Reviewer: ```suggestion
The Salesforce streaming Inbound Endpoint allows you to perform various operations on Salesforce streaming data via WSO2 EI.
```

Author: Updated.

> The Salesforce streaming API receives notifications based on the changes that happen to Salesforce data with respect to an SQQL (Salesforce Object Query Language) query you define, in a secured and scalable way. For more information, navigate to [Salesforce streaming documentation](https://developer.salesforce.com/docs/atlas.en-us.202.0.api_streaming.meta/api_streaming/quick_start_workbench.htm).

Reviewer: Shall we add a link to the API if possible?

Author: Added Streaming API Developer Guide to the doc.

## What you'll build

> The Salesforce inbound endpoint supported via the WSO2 EI is a listening inbound endpoint that can consume messages from the Salesforce. Then injects messages to a sequence in WSO2 EI. However, for simplicity of this example, we will just log the message. You can extend the sample as required using WSO2 [EI mediators](https://docs.wso2.com/display/EI660/ESB+Mediators). 

Reviewer: ```suggestion
The Salesforce inbound endpoint supported via the WSO2 EI is a listening inbound endpoint that can consume messages from Salesforce. This injects messages to a sequence in WSO2 EI. However, for simplicity of this example, we will just log the message. You can extend the sample as required using WSO2 [EI mediators](https://docs.wso2.com/display/EI660/ESB+Mediators). 
```

Let's change this link to a relative link in the 7.0.0 docs.

Author: Updated.

> The Salesforce inbound endpoint supported via the WSO2 EI is a listening inbound endpoint that can consume messages from Salesforce. This injects messages to a sequence in WSO2 EI. However, for simplicity of this example, we will just log the message. You can extend the sample as required using WSO2 [EI mediators](https://ei.docs.wso2.com/en/latest/micro-integrator/references/mediators/about-mediators/). 

Reviewer: Let's use a relative link here. Otherwise we have to manually update this when we release a new version of the documentation.

Author: Updated as ../../references/mediators/about-mediators/

> In this exmple we can trigger the notifications to the Salesforce Inbound Endpoint via creating the `Platform events` or `PushTopic` methods. Please note that our example configurations are based on the creating `PushTopic` method. You can use the instructions given in the [sf-rest inbound endpoint configuration](sf-rest inbound endpoint configuration.md) documentation.

Reviewer: ```suggestion
In this exmple we can trigger the notifications to the Salesforce Inbound Endpoint via creating the `Platform events` or `PushTopic` methods. Please note that our example configurations are based on creating the `PushTopic` method. You can use the instructions given in the [sf-rest inbound endpoint configuration](sf-rest inbound endpoint configuration.md) documentation.
```

Author: Updated.

The following diagram illustrates all the required functionality of the Salesforce inbound operations that you are going to build. 

> For example, we are building integrated example driven through the [Salesforce connector](sf-rest-connector-example.md) and Salesforce Inbound Endpoint.The user calls the Salesforcerest API. It invoke the **create** sequence and create a new account in Salesforce. Then through the **retrieve** sequence, it displays all the existing account details to the user. 

Reviewer: ```suggestion
For example, we are building integrated example driven through the [Salesforce connector](sf-rest-connector-example.md) and Salesforce Inbound Endpoint. The user calls the Salesforcerest API. It invokes the **create** sequence and creates a new account in Salesforce. Then through the **retrieve** sequence, it displays all the existing account details to the user. 
```

Author: Updated.

> That you have configured the Salesforce Inbound Endpoint, use the following Inbound Endpoint configuration to retrieve account details from your Salesforce account. The Salesforce inbound endpoint of WSO2 EI acts as a message receiver. You can inject that message into the mediation flow for getting the required output.enen

Reviewer: ```suggestion
Now that you have configured the Salesforce Inbound Endpoint, use the following Inbound Endpoint configuration to retrieve account details from your Salesforce account. The Salesforce inbound endpoint of WSO2 EI acts as a message receiver. You can inject that message into the mediation flow for getting the required output.
```

Author: Updated.

<img src="../../../../assets/img/connectors/salesforce-inboundEP-example.png" title="Salesforce Inbound Endpoint" width="800" alt="Salesforce Inbound Endpoint"/>

## Configure inbound endpoint using WSO2 Integration Studio

1. Download [WSO2 Integration Studio](https://wso2.com/integration/integration-studio/). Create an **Integration Project** as below. 

   <img src="../../../../assets/img/connectors/integration-project.png" title="Creating a new Integration Project" width="800" alt="Creating a new Integration Project" />

> 2. Right click on **Source** -> **main** -> **synapse-config** -> **inbound-endpoints** and add a new **custom inbound endpoint**.

Reviewer: Shall we use exact terms from the UI?

Author: Updated.

         <parameter name="connection.salesforce.packageVersion">37.0</parameter>
         <parameter name="connection.salesforce.salesforceObject">/topic/Account</parameter>
         <parameter name="connection.salesforce.loginEndpoint">https://login.salesforce.com</parameter>
         <parameter name="connection.salesforce.userName">Username</parameter>
         <parameter name="connection.salesforce.password">test123XXXXXXXXXX</parameter>
         <parameter name="connection.salesforce.waitTime">5000</parameter>
         <parameter name="connection.salesforce.connectionTimeout">20000</parameter>
         <parameter name="connection.salesforce.soapApiVersion">22.0</parameter>
      </parameters>
   </inboundEndpoint>
   ```
   Sequence to process the message.
   
   In this example for simplicity we will just log the message, but in a real world use case, this can be any type of message mediation.
 
   ```xml
   <?xml version="1.0" encoding="UTF-8"?>
   <sequence name="test" onError="fault" xmlns="http://ws.apache.org/ns/synapse">
       <log level="full"/>
       <drop/>
   </sequence>
   ```

> > **Note**: To configure the `connection.salesforce.password` parameter value please use the steps given under topic `Reset Security Token` in the [sf-rest inbound endpoint configuration](sf-rest inbound endpoint configuration.md) document.

Reviewer: ```suggestion
> **Note**: To configure the `connection.salesforce.password` parameter value, please use the steps given under the topic `Reset Security Token` in the [sf-rest inbound endpoint configuration](sf-rest inbound endpoint configuration.md) document.
```

Does this link work? File name normally doesn't have spaces.

Author: Updated.

1. Right click on Composite Application Project and click on **Export Composite Application Project**.</br> 
  <img src="../../../../assets/img/connectors/capp-project1.png" title="Export as a Carbon Application" width="300" alt="Export as a Carbon Application" />

2. Select an **Export Destination** where you want to save the .car file. 

3. In the next **Create a deployable CAR file** screen, select inbound endpoint and sequence artifacts and click **Finish**. The CApp will get created at the specified location provided in the previous step. 

## Deployment

1. Navigate to the [connector store](https://store.wso2.com/store/assets/esbconnector/list) and search for `SalesforceRest`. Click on `Salesforce Inbound Endpoint` and download the .jar file by clicking on `Download Inbound Endpoint`. Copy this .jar file into  <PRODUCT-HOME>/lib folder. 

2. Copy the exported carbon application to the <PRODUCT-HOME>/repository/deployment/server/carbonapps folder. 

4. [Start the WSO2 EI server](https://ei.docs.wso2.com/en/latest/micro-integrator/overview/quick-start-guide/#start-the-micro-integrator). 

## Testing  

> > **Note**: If you want to test this scenario using insert data manually into the created object records please follow the steps given under topic `Testing the PushTopic Channel` in the [sf-rest inbound endpoint configuration](sf-rest inbound endpoint configuration.md) document.

Reviewer: ```suggestion
> **Note**: If you want to test this scenario by inserting data manually into the created object records, please follow the steps given under topic `Testing the PushTopic Channel` in the [sf-rest inbound endpoint configuration](sf-rest inbound endpoint configuration.md) document.
```

Is this link correct?

Author: Updated.

   Please use the [sf-rest-connector-example](sf-rest-connector-example.md) testing steps to test this Inbound Endpoint scenario;
   

>    Save a file called data.json with the following payload. (Change the value of `Name` field as `Manager` )

Reviewer: ```suggestion
   Save a file called data.json with the following payload (change the value of `Name` field as `Manager`).
```

Author: Updated.

   ```
   {
   	"sObject":"Account",
   	"fieldAndValue": {
       "name": "Manager",
       "description":"This Account belongs to WSO2"
     }
   }
   ``` 
   Invoke the API as shown below using the curl command. Curl application can be downloaded from [here](https://curl.haxx.se/download.html).   
   
   ```
   curl -X POST -d @data.json http://localhost:8280/salesforcerest --header "Content-Type:application/json"
   ```

>    You will get a set of account names and the respective IDs as the output. At the same time in WSO2 EI console you can see the following message.

Reviewer: ```suggestion
   You will get a set of account names and the respective IDs as the output. At the same time, in the WSO2 EI console you can see the following message.
```

Author: Updated.
